pyo3-macros-backend/src/attributes.rs

```diff
@@ -134,28 +131,156 @@ pub fn is_attribute_ident(attr: &syn::Attribute, name: &str) -> bool {
 /// (In `retain`, returning `true` keeps the element, here it removes it.)
 pub fn take_attributes(
     attrs: &mut Vec<Attribute>,
-    mut extractor: impl FnMut(&Attribute) -> Result<bool>,
+    mut extractor: impl FnMut(&mut Attribute) -> Result<bool>,
 ) -> Result<()> {
     *attrs = attrs
         .drain(..)
-        .filter_map(|attr| {
-            extractor(&attr)
+        .filter_map(|mut attr| {
+            extractor(&mut attr)
                 .map(move |attribute_handled| if attribute_handled { None } else { Some(attr) })
                 .transpose()
         })
         .collect::<Result<_>>()?;
     Ok(())
 }
 
-pub fn take_pyo3_options<T: Parse>(attrs: &mut Vec<syn::Attribute>) -> Result<Vec<T>> {
+pub fn take_pyo3_options<T: Parse>(attrs: &mut Vec<Attribute>) -> Result<Vec<T>> {
     let mut out = Vec::new();
-    take_attributes(attrs, |attr| {
-        if let Some(options) = get_pyo3_options(attr)? {
+    let mut new_attrs = Vec::new();
+
+    for mut attr in attrs.drain(..) {
+        let parse_attr = |meta, _attributes: &Attribute| {
+            if let Meta::List(meta_list) = meta {
+                if meta_list.path.is_ident("pyo3") {
+                    let parsed = Punctuated::<_, Token![,]>::parse_terminated
+                        .parse2(meta_list.nested.to_token_stream())?;
+                    out.extend(parsed.into_iter());
+                    return Ok(true);
+                }
+            }
+            Ok(false)
+        };
+        if let Ok(mut meta) = attr.parse_meta() {
+            if handle_cfg_feature_pyo3(&mut attr, &mut meta, parse_attr)? {
+                continue;
+            }
+        }
```

Comment:
I think it would be good to check `if attr.path.is_ident("cfg_attr")` before calling `parse_meta`, to avoid a lot of unnecessary parsing work.

Reply:
Is parsing `Meta` expensive? This is the only location where we do it in this order, in the other two locations we also need `meta` for the normal parsing

Reply:
So I've now written a benchmark in #2794 

It's based against main, what I see on my machine is that the cost per attribute (for a very basic attribute) goes from about 250ns to 1us. Most of that overhead is removed again by just adding this check.

So sure, while it's not much, maybe in a bigger codebase with lots of attributes this could eventually add up to a second or so. I'd like us to try to keep macro performance as good as possible, especially when it's so easy to add the check here.

(Generally I've been heading in the direction of avoiding `meta` parsing and writing more specific parse implementations like the one here, for best efficiency.)

Reply:
sweet, i didn't expect this much rigor! i've now gated both parse with `is_ident`

Reply:
fwiw, i think moving everything into `pyo3(...)` would be help because then we can skip the non-pyo3 macro parsing completely for attributes that are not ours